File: src/main/java/com/gemini/krakenbot/service/impl/KrakenServiceImpl.java
package com.gemini.krakenbot.service.impl;

import com.gemini.krakenbot.service.ConfigService;


import lombok.extern.slf4j.Slf4j;


import com.gemini.krakenbot.service.KrakenService;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;

import org.springframework.stereotype.Service;
import org.springframework.web.client.RestClient;

import javax.crypto.Mac;
import javax.crypto.spec.SecretKeySpec;
import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.util.Base64;
import java.util.HashMap;
import java.util.Map;
import java.util.stream.Collectors;
import java.util.concurrent.atomic.AtomicLong;

@Service
@Slf4j
public class KrakenServiceImpl implements KrakenService {
    private static final String API_URL = "https://api.kraken.com";
    private static final String API_VERSION = "0";

    private final RestClient restClient;
    private final ConfigService configService;
    private final ObjectMapper objectMapper;
    private final AtomicLong nonceGenerator = new AtomicLong(System.currentTimeMillis() * 1000);

    public KrakenServiceImpl(ConfigService configService, ObjectMapper objectMapper, RestClient.Builder restClientBuilder) {
        this.configService = configService;
        this.objectMapper = objectMapper;
        this.restClient = restClientBuilder.baseUrl(API_URL).build();
    }

    public Map<String, Double> getBalances() {
        String path = "/" + API_VERSION + "/private/Balance";
        JsonNode response = queryPrivate(path, Map.of());
        Map<String, Double> balances = new HashMap<>();
        response.properties().forEach(entry -> balances.put(entry.getKey(), entry.getValue().asDouble()));
        return balances;
    }

    public Map<String, Double> getTickerPrices(String pairs) {
        String path = "/" + API_VERSION + "/public/Ticker?pair=" + pairs;
        JsonNode result = queryPublic(path).path("result");
        Map<String, Double> prices = new HashMap<>();
        result.properties().forEach(entry -> {
            // Kraken returns array [price, ...], usually index 0 of 'c' (last trade closed)
            // is good enough,
            // or 'c'[0]. Let's check structure. Typically
            // {"XXBTZUSD":{"a":["..."],"b":["..."],"c":["63000.0","0.1"]}}
            // We want current price, let's use 'c'[0] (last trade price).
            JsonNode c = entry.getValue().path("c");
            if (c.isArray() && !c.isEmpty()) {
                prices.put(entry.getKey(), c.get(0).asDouble());
            }
        });
        return prices;
    }

    public void executeOrder(String pair, String type, String side, double volume) {
        if (configService.getConfig().settings().dryRun()) {
            log.info("[DRY RUN] Would execute order: {} {} {} volume={}", type, side, pair, volume);
            return;
        }

        String path = "/" + API_VERSION + "/private/AddOrder";
        Map<String, String> params = new HashMap<>();
        params.put("pair", pair);
        params.put("type", side); // buy/sell
        params.put("ordertype", type); // market/limit
        params.put("volume", String.valueOf(volume));

        try {
            JsonNode resp = queryPrivate(path, params);
            log.info("Order Executed: {}", resp.toString());
        } catch (Exception e) {
            log.error("Failed to execute order: {} {} {} volume={}", type, side, pair, volume, e);
        }
    }

    @SuppressWarnings("null")
    private JsonNode queryPublic(String path) {
        String response = restClient.get().uri(path).retrieve().body(String.class);
        if (response == null)
            response = "{}";
        try {
            JsonNode root = objectMapper.readTree(response);
            if (root.has("error") && !root.path("error").isEmpty()) {
                log.error("Kraken Public API Error for path {}: {}", path, root.path("error"));
                // Return empty result or throw? Throwing allows catching in main loop.
                throw new RuntimeException("Kraken Public API Error: " + root.path("error").toString());
            }
            return root;
        } catch (JsonProcessingException e) {
            throw new RuntimeException("Failed to parse public API response", e);
        }
    }

    @SuppressWarnings("null")
    private JsonNode queryPrivate(String path, Map<String, String> data) {
        int maxRetries = 5;
        int retryCount = 0;

        while (true) {
            String nonce = String.valueOf(nonceGenerator.incrementAndGet());
            Map<String, String> payload = new HashMap<>(data);
            payload.put("nonce", nonce);

            String postData = payload.entrySet().stream()
                    .map(e -> e.getKey() + "=" + e.getValue())
                    .collect(Collectors.joining("&"));

            String signature = signRequest(path, nonce, postData);

            String apiKey = configService.getConfig().kraken().apiKey();
            if (apiKey == null)
                throw new RuntimeException("API Key is null");

            String response = restClient.post()
                    .uri(path)
                    .header("API-Key", apiKey)
                    .header("API-Sign", signature)
                    .header("Content-Type", "application/x-www-form-urlencoded")
                    .body(postData)
                    .retrieve()
                    .body(String.class);

            if (response == null)
                response = "{}";

            try {
                JsonNode root = objectMapper.readTree(response);
                if (!root.path("error").isEmpty()) {
                    String errorMsg = root.path("error").toString();
                    if (errorMsg.contains("Invalid nonce") && retryCount < maxRetries) {
                        log.warn("Invalid nonce detected. Adjusting nonce generator and retrying (Attempt {}/{})", retryCount + 1, maxRetries);
                        nonceGenerator.addAndGet(5000); // jump ahead to resolve collisions
                        retryCount++;
                        continue;
                    }
                    throw new RuntimeException("Kraken API Error: " + errorMsg);
                }
                return root.path("result");
            } catch (JsonProcessingException e) {
                throw new RuntimeException("Failed to parse private API response", e);
            }
        }
    }

    private String signRequest(String path, String nonce, String postData) {
        try {
            byte[] sha2 = MessageDigest.getInstance("SHA-256")
                    .digest((nonce + postData).getBytes(StandardCharsets.UTF_8));

            byte[] pathBytes = path.getBytes(StandardCharsets.UTF_8);
            byte[] hmacMessage = new byte[pathBytes.length + sha2.length];
            System.arraycopy(pathBytes, 0, hmacMessage, 0, pathBytes.length);
            System.arraycopy(sha2, 0, hmacMessage, pathBytes.length, sha2.length);

            Mac mac = Mac.getInstance("HmacSHA512");
            byte[] secretDecoded = Base64.getDecoder().decode(configService.getConfig().kraken().privateKey());
            SecretKeySpec secretSpec = new SecretKeySpec(secretDecoded, "HmacSHA512");
            mac.init(secretSpec);

            byte[] sigBytes = mac.doFinal(hmacMessage);
            return Base64.getEncoder().encodeToString(sigBytes);
        } catch (Exception e) {
            throw new RuntimeException("Failed to sign request", e);
        }
    }
}

File: src/main/kotlin/com/gemini/krakenbot/service/impl/KrakenServiceImpl.kt
package com.gemini.krakenbot.service.impl

import com.fasterxml.jackson.databind.JsonNode
import com.fasterxml.jackson.databind.ObjectMapper
import com.gemini.krakenbot.domain.OrderResult
import com.gemini.krakenbot.domain.RawBalances
import com.gemini.krakenbot.domain.RawPrices
import com.gemini.krakenbot.model.KrakenApiConstants
import com.gemini.krakenbot.model.LedgerEvent
import com.gemini.krakenbot.model.TradeRecord
import com.gemini.krakenbot.service.BoundedTradeHistoryService
import com.gemini.krakenbot.service.ConfigService
import com.gemini.krakenbot.service.KrakenService
import com.gemini.krakenbot.service.SpendableBalanceService
import com.gemini.krakenbot.util.PrecisionConstants
import io.ktor.client.HttpClient
import io.ktor.client.plugins.ResponseException
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.flow.flow
import org.slf4j.LoggerFactory
import java.io.IOException
import java.math.BigDecimal
import java.math.RoundingMode
import java.util.concurrent.atomic.AtomicInteger
import java.util.concurrent.atomic.AtomicLong
import kotlin.time.Duration.Companion.milliseconds
import kotlin.time.Duration.Companion.minutes

class KrakenServiceImpl(
    private val configService: ConfigService,
    private val objectMapper: ObjectMapper,
    private val httpClient: HttpClient,
    private val rateLimiter: RateLimiter = RateLimiter(),
) : KrakenService,
    SpendableBalanceService,
    BoundedTradeHistoryService {
    private val log = LoggerFactory.getLogger(KrakenServiceImpl::class.java)

    private val nonceGenerator = AtomicLong(System.currentTimeMillis() * 1_000_000L)

    /** Bounds exception text persisted into order error rows / dashboard payloads. */
    private companion object {
        const val MAX_ERROR_MESSAGE_LENGTH = 500
    }

    private val transport = KrakenTransport(
        configService = configService,
        objectMapper = objectMapper,
        httpClient = httpClient,
        rateLimiter = rateLimiter,
        nonceGenerator = nonceGenerator,
    )

    private val lastFetchedCount = AtomicInteger(0)
    private val lastLedgerCount = AtomicInteger(0)

    override fun getLastTradeHistoryTotalCount(): Int = lastFetchedCount.get()

    override fun getLastLedgerTotalCount(): Int = lastLedgerCount.get()

    override suspend fun getApiCallCounter(): Double = rateLimiter.getCurrentCounter()

    private suspend fun <T> retryWithFlow(
        actionName: String,
        maxAttempts: Int = 5,
        maxLockoutAttempts: Int = 9,
        initialBackoffMs: Long = 2000,
        rateLimitBackoffMs: Long = 10000,
        initialLockoutBackoffMs: Long = 10_000,
        maxLockoutBackoffMs: Long = 15.minutes.inWholeMilliseconds,
        block: suspend () -> T,
    ): T = flow {
        var currentBackoff = initialBackoffMs
        var currentRateLimitBackoff = rateLimitBackoffMs
        var currentLockoutBackoff = initialLockoutBackoffMs
        var attempt = 0
        var lockoutAttempt = 0

        while (true) {
            try {
                emit(block())
                return@flow
            } catch (e: CancellationException) {
                throw e
            } catch (e: Exception) {
                val status = (e as? ResponseException)?.response?.status?.value
                val isRawRateLimit = status == 429
                val isRawLockout = status == 503
                val isRateLimit =
                    isRawRateLimit || e.message?.contains(KrakenApiConstants.ERROR_RATE_LIMIT_EXCEEDED) == true
                val isLockout = isRawLockout || e.message?.contains(KrakenApiConstants.ERROR_TEMPORARY_LOCKOUT) == true
                val isNetworkOrTransient = e is IOException || e is ResponseException
                val retryable = isNetworkOrTransient || isRateLimit || isLockout
                val attemptsUsed = if (isLockout) lockoutAttempt else attempt
                val attemptLimit = if (isLockout) maxLockoutAttempts else maxAttempts

                if (retryable && attemptsUsed < attemptLimit - 1) {
                    val waitTime =
                        when {
                            isLockout -> currentLockoutBackoff.coerceAtMost(maxLockoutBackoffMs)
                            isRateLimit -> currentRateLimitBackoff
                            else -> currentBackoff
                        }
                    log.warn(
                        "Transient failure in {} (attempt {}/{}). Retrying in {}ms... Error: {}",
                        actionName,
                        attemptsUsed + 1,
                        attemptLimit,
                        waitTime,
                        e.message,
                    )
                    delay(waitTime.milliseconds)

                    when {
                        isLockout -> {
                            currentLockoutBackoff =
                                (currentLockoutBackoff * 2).coerceAtMost(maxLockoutBackoffMs)
                            lockoutAttempt++
                        }

                        isRateLimit -> {
                            currentRateLimitBackoff *= 2
                            attempt++
                        }

                        else -> {
                            currentBackoff *= 2
                            attempt++
                        }
                    }
                } else {
                    throw e
                }
            }
        }
    }.first()

    override suspend fun getBalances(): RawBalances {
        val path = KrakenApiConstants.PATH_BALANCE
        val response = queryPrivate(path, emptyMap())
        return KrakenParsers.parseBalances(response)
    }

    override suspend fun getSpendableBalances(): RawBalances {
        val response = queryPrivate(KrakenApiConstants.PATH_BALANCE_EX, emptyMap())
        return KrakenParsers.parseSpendableBalances(response)
    }

    override suspend fun getTickerPrices(pairs: String): RawPrices {
        val path = "${KrakenApiConstants.PATH_TICKER}?${KrakenApiConstants.PARAM_PAIR}=$pairs"
        val result = queryPublic(path).path(KrakenApiConstants.FIELD_RESULT)
        return KrakenParsers.parseTickerPrices(result)
    }

    override suspend fun executeOrder(
        pair: String,
        type: String,
        side: String,
        volume: BigDecimal,
        dryRun: Boolean?,
        clOrdId: String?,
    ): OrderResult {
        val normalizedVolume =
            volume
                .setScale(
                    PrecisionConstants.SCALE_CRYPTO,
                    RoundingMode.DOWN,
                ).stripTrailingZeros()

        val isDryRun = dryRun ?: configService.getConfig().settings.dryRun.also {
            log.warn("executeOrder called without a dryRun argument; resolved from live config: {}", it)
        }
        if (isDryRun) {
            log.info(
                "[DRY RUN] Would execute order: {} {} {} volume={} cl_ord_id={}",
                type,
                side,
                pair,
                normalizedVolume.toPlainString(),
                clOrdId,
            )
            return OrderResult(
                success = true,
                pair = pair,
                side = side,
                volume = normalizedVolume,
                dryRun = true,
            )
        }

        val path = KrakenApiConstants.PATH_ADD_ORDER
        val params =
            mutableMapOf(
                KrakenApiConstants.PARAM_PAIR to pair,
                KrakenApiConstants.PARAM_TYPE to side,
                KrakenApiConstants.PARAM_ORDERTYPE to type,
                KrakenApiConstants.PARAM_VOLUME to normalizedVolume.toPlainString(),
            )
        if (clOrdId != null) {
            params[KrakenApiConstants.PARAM_CL_ORD_ID] = clOrdId
        }

        return try {
            val resp = queryPrivate(path, params)
            val txidNode = resp.path(KrakenApiConstants.FIELD_TXID)
            val orderTxid =
                if (txidNode.isArray && txidNode.size() > 0) {
                    txidNode[0].asText().ifBlank { null }
                } else {
                    null
                }
            log.info(
                "Order executed pair={} side={} volume={} txid={}",
                pair,
                side,
                normalizedVolume.toPlainString(),
                orderTxid,
            )
            if (orderTxid == null) {
                OrderResult(
                    success = false,
                    pair = pair,
                    side = side,
                    volume = normalizedVolume,
                    errorMessage = "Kraken AddOrder response did not contain a transaction id",
                    submissionUncertain = true,
                )
            } else {
                OrderResult(
                    success = true,
                    pair = pair,
                    side = side,
                    volume = normalizedVolume,
                    orderTxid = orderTxid,
                )
            }
        } catch (e: CancellationException) {
            throw e
        } catch (e: Exception) {
            val message = e.message.orEmpty().ifEmpty { e.javaClass.simpleName }.take(MAX_ERROR_MESSAGE_LENGTH)
            log.error(
                "Failed to execute order: {} {} {} volume={}",
                type,
                side,
                pair,
                normalizedVolume.toPlainString(),
                e,
            )
            OrderResult(
                success = false,
                pair = pair,
                side = side,
                volume = normalizedVolume,
                errorMessage = message,
                submissionUncertain = isAmbiguousSubmissionFailure(e),
            )
        }
    }

    private fun isAmbiguousSubmissionFailure(error: Throwable): Boolean = generateSequence(error) { it.cause }
        .any { cause ->
            cause is AmbiguousOrderSubmissionException ||
                cause is IOException ||
                cause is ResponseException
        }

    override suspend fun getTradeHistory(startSec: Long?, offset: Int?): List<TradeRecord> =
        getTradeHistoryUntil(startSec, offset, null)

    override suspend fun getTradeHistoryUntil(startSec: Long?, offset: Int?, endSec: Long?): List<TradeRecord> {
        if (!configService.getConfig().kraken.hasValidCredentials()) {
            log.warn("Kraken API key is blank or placeholder. Skipping trade history fetch.")
            return emptyList()
        }

        val params = mutableMapOf<String, String>()
        if (startSec != null) params[KrakenApiConstants.PARAM_START] = startSec.toString()
        if (endSec != null) params[KrakenApiConstants.PARAM_END] = endSec.toString()
        if (offset != null) params[KrakenApiConstants.PARAM_OFS] = offset.toString()

        val result =
            try {
                queryPrivate(KrakenApiConstants.PATH_TRADES_HISTORY, params)
            } catch (e: CancellationException) {
                throw e
            } catch (e: Exception) {
                log.error("Failed to query private TradesHistory endpoint", e)
                throw e
            }

        val allocations = configService.getConfig().allocations.map { it.symbol.value }
        val (trades, count) = KrakenParsers.parseTradeHistory(result, allocations)
        lastFetchedCount.set(count)
        return trades
    }

    override suspend fun getLedgers(
        startSec: Long?,
        offset: Int?,
        endSec: Long?,
        types: Set<String>?,
    ): List<LedgerEvent> {
        if (!configService.getConfig().kraken.hasValidCredentials()) {
            log.warn("Kraken API key is blank or placeholder. Skipping ledger fetch.")
            return emptyList()
        }

        val params = mutableMapOf<String, String>()
        if (startSec != null) params[KrakenApiConstants.PARAM_START] = startSec.toString()
        if (endSec != null) params[KrakenApiConstants.PARAM_END] = endSec.toString()
        if (offset != null) params[KrakenApiConstants.PARAM_OFS] = offset.toString()

        val sortedTypes = types?.sorted()
        if (sortedTypes != null && sortedTypes.size > 1) {
            val fanOutParams = params - KrakenApiConstants.PARAM_OFS
            val pages = sortedTypes.map { type ->
                queryLedgerPage(fanOutParams + (KrakenApiConstants.PARAM_TYPE to type), types)
            }
            lastLedgerCount.set(pages.sumOf { it.second })
            return pages.flatMap { it.first }
        }
        val pageParams = if (sortedTypes != null) {
            params + (KrakenApiConstants.PARAM_TYPE to sortedTypes.single())
        } else {
            params
        }
        val (entries, count) = queryLedgerPage(pageParams, types)
        lastLedgerCount.set(count)
        return entries
    }

    private suspend fun queryLedgerPage(
        params: Map<String, String>,
        expectedTypes: Set<String>?,
    ): Pair<List<LedgerEvent>, Int> {
        val result =
            try {
                queryPrivate(KrakenApiConstants.PATH_LEDGERS, params)
            } catch (e: CancellationException) {
                throw e
            } catch (e: Exception) {
                log.error("Failed to query private Ledgers endpoint", e)
                throw e
            }

        return KrakenParsers.parseLedgerPage(result, expectedTypes)
    }

    override suspend fun getOHLC(pair: String, interval: Int, since: Long?): List<Pair<Long, BigDecimal>> {
        val params = mutableMapOf<String, String>()
        params[KrakenApiConstants.PARAM_PAIR] = pair
        params[KrakenApiConstants.PARAM_INTERVAL] = interval.toString()
        if (since != null) params[KrakenApiConstants.PARAM_SINCE] = since.toString()
        val queryStr = params.map { "${it.key}=${it.value}" }.joinToString("&")
        val path = "${KrakenApiConstants.PATH_OHLC}?$queryStr"
        val root =
            try {
                queryPublic(path)
            } catch (e: CancellationException) {
                throw e
            } catch (e: Exception) {
                log.error("Failed to query public OHLC endpoint for pair $pair", e)
                return emptyList()
            }

        return KrakenParsers.parseOHLC(root, pair)
    }

    private suspend fun queryPublic(path: String): JsonNode = retryWithFlow("queryPublic($path)") {
        transport.queryPublic(path)
    }

    private suspend fun queryPrivate(path: String, data: Map<String, String>): JsonNode = retryWithFlow(
        actionName = "queryPrivate($path)",
        maxAttempts = if (path == KrakenApiConstants.PATH_ADD_ORDER) 1 else 5,
        maxLockoutAttempts = if (path == KrakenApiConstants.PATH_ADD_ORDER) 1 else 9,
    ) {
        transport.queryPrivate(path, data)
    }
}
